File: solution/day19/src/test/java/blog/ArticleTests.java
package blog;

import io.vavr.control.Either;
import org.assertj.vavr.api.VavrAssertions;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.function.Consumer;
import java.util.function.Function;

import static blog.ArticleBuilder.*;
import static org.assertj.core.api.Assertions.assertThat;
import static org.instancio.Instancio.create;

class ArticleTests {
    private Either<Error, Article> result;

    @Test
    void should_add_comment_in_an_article() {
        when(article -> article.addComment(COMMENT_TEXT, AUTHOR));
        then(result -> VavrAssertions.assertThat(result).hasRightValueSatisfying(article -> {
            assertComment(article.getComments().last(), COMMENT_TEXT, AUTHOR);
        }));
    }

    @Test
    void should_add_comment_in_an_article_containing_already_a_comment() {
        final var newComment = create(String.class);
        final var newAuthor = create(String.class);

        when(ArticleBuilder::commented, article -> article.addComment(newComment, newAuthor));
        then(result -> VavrAssertions.assertThat(result).hasRightValueSatisfying(article -> {
            assertThat(article.getComments()).hasSize(2);
            assertComment(article.getComments().last(), newComment, newAuthor);
        }));
    }

    @Nested
    class Fail {
        @Test
        void when_adding_an_existing_comment() {
            when(ArticleBuilder::commented, article -> article.addComment(article.getComments().get(0).text(), article.getComments().get(0).author()));
            then(result -> VavrAssertions.assertThat(result)
                    .containsOnLeft(new Error("This comment already exists in this article")))
            ;
        }
    }

    private static void assertComment(Comment comment, String commentText, String author) {
        assertThat(comment.text()).isEqualTo(commentText);
        assertThat(comment.author()).isEqualTo(author);
    }

    private void when(ArticleBuilder articleBuilder, Function<Article, Either<Error, Article>> act) {
        result = act.apply(
                articleBuilder.build()
        );
    }

    private void when(Function<Article, Either<Error, Article>> act) {
        when(anArticle(), act);
    }

    private void when(Function<ArticleBuilder, ArticleBuilder> options, Function<Article, Either<Error, Article>> act) {
        when(options.apply(anArticle()), act);
    }

    private void then(Consumer<Either<Error, Article>> act) {
        act.accept(result);
    }
}

File: solution/c#/Day19/Day19.Tests/ArticleTests.cs
using FluentAssertions;
using FluentAssertions.LanguageExt;
using LanguageExt;
using LanguageExt.UnsafeValueAccess;
using Xunit;
using static Day19.Tests.ArticleBuilder;

namespace Day19.Tests
{
    public class ArticleTests
    {
        private Either<Error, Article> _result;
        private readonly Bogus.Randomizer _random = new();

        [Fact]
        public void Should_Add_Comment_In_An_Article()
        {
            Given(AnArticle());
            When(article => article.AddComment(CommentText, Author));
            Then(result
                => result.Should()
                    .BeRight(article =>
                        {
                            article.Comments.Should().HaveCount(1);
                            AssertComment(article.Comments[0], CommentText, Author);
                        }
                    ));
        }

        [Fact]
        public void Should_Add_Comment_In_An_Article_Containing_Already_A_Comment()
        {
            var newComment = _random.String(10);
            var newAuthor = _random.String(3);

            Given(AnArticle().Commented());
            When(article => article.AddComment(newComment, newAuthor));
            Then(result
                => result.Should()
                    .BeRight(article =>
                    {
                        article.Comments.Should().HaveCount(2);
                        AssertComment(article.Comments[1], newComment, newAuthor);
                    })
            );
        }

        private static void AssertComment(Comment comment, string expectedComment, string expectedAuthor)
        {
            comment.Text.Should().Be(expectedComment);
            comment.Author.Should().Be(expectedAuthor);
            comment.CreationDate.Should().Be(DateOnly.FromDateTime(DateTime.Now));
        }

        public class Fail : ArticleTests
        {
            [Fact]
            public void When_Adding_An_Existing_Comment()
            {
                Given(AnArticle().Commented());
                When(article => article.AddComment(CommentText, Author));
                Then(result
                    => result.Should()
                        .BeLeft(error =>
                            error.Should()
                                .Be(new Error("This comment already exists in this article"))
                        ));
            }
        }

        private void Given(ArticleBuilder articleBuilder) => _result = articleBuilder.Build();
        private void When(Func<Article, Either<Error, Article>> act) => _result = act(_result.ValueUnsafe());
        private void Then(Action<Either<Error, Article>> act) => act(_result);
    }
}